lib/attrsets.nix

```
@@ -2137,6 +2137,6 @@ rec {
     "lib.zip is a deprecated alias of lib.zipAttrsWith." zipAttrsWith;
 
   # DEPRECATED
-  cartesianProductOfSets = warnIf (isInOldestRelease 2405)
+  cartesianProductOfSets = warnIf (oldestSupportedReleaseIsAtLeast 2405)
     "lib.cartesianProductOfSets is a deprecated alias of lib.cartesianProduct." cartesianProduct;
```

This is maybe a _bit_ too long, but I think it's a much better description:

```suggestion
  cartesianProductOfSets = warnIf (allSupportedReleasesSupportAlternativeOriginallyIntroducedIn 2405)
    "lib.cartesianProductOfSets is a deprecated alias of lib.cartesianProduct." cartesianProduct;
```

Hard to remove parts without distorting the meaning though :/

---

The difference is that this description includes the policy, whereas mine only describes the mechanism. I think that's a sensible rule of thumb.

The policy should be documented, certainly, but maybe not in the _name_ of the function.

Also, more high level level functions are possible, such as
```nix
foo =
  deprecateForAlternativeIn
    2405
    "use bar instead of foo"
    bar;
# or similar to `mkRenamedOptionModule` iirc, though could be misread
foo = deprecate { old = "package foo"; new = "bar"; sinceRelease 2405; } bar;
```

Also, such a function could both warn and throw if we decide a standard time frame for that.

Anyway, before veering too far off topic, I think we could do with a slightly better named low level function first.
It will save a few brain cycles when developing the higher level ones.

---

Makes sense and again reminds me of NixOS/rfcs#33 :)